src/Compiler/Microsoft.AspNetCore.Razor.Language/src/DefaultBoundAttributeDescriptorBuilder.cs

```diff
@@ -1,80 +1,73 @@
 ﻿// Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
-#nullable disable
-
 using System;
 using System.Collections.Generic;
-using System.Linq;
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using Microsoft.AspNetCore.Razor.PooledObjects;
+using Microsoft.Extensions.ObjectPool;
 
 namespace Microsoft.AspNetCore.Razor.Language;
 
-internal class DefaultBoundAttributeDescriptorBuilder : BoundAttributeDescriptorBuilder
+internal class DefaultBoundAttributeDescriptorBuilder : BoundAttributeDescriptorBuilder, IBuilder<BoundAttributeDescriptor>
 {
-    private static readonly IReadOnlyDictionary<string, string> PrimitiveDisplayTypeNameLookups = new Dictionary<string, string>(StringComparer.Ordinal)
+    private static readonly ObjectPool<HashSet<BoundAttributeParameterDescriptor>> s_boundAttributeParameterSetPool
+        = HashSetPool<BoundAttributeParameterDescriptor>.Create(BoundAttributeParameterDescriptorComparer.Default);
+
+    private static readonly ImmutableDictionary<string, string> s_primitiveDisplayTypeNameLookups = new[]
```

Comment:
📝 `ImmutableDictionary<TKey, TValue>` is much less efficient than `Dictionary<TKey, TValue>` for pure reads. If this set is never modified, consider using `ImmutableSegmentedDictionary<TKey, TValue>` from Roslyn instead.

Reply:
Unfortunately, we don't have access to `ImmutableSegmentedDictionary<TKey, TValue>` and would need to maintain another copy in Razor. I'll switch this back to a `Dictionary<TKey, TValue>` for faster lookup with a comment.